// @liquidmetal-ai/lm-raindrop/resources/query/query.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as EpisodicMemoryAPI from "./episodic-memory.mjs";
import { EpisodicMemory } from "./episodic-memory.mjs";
import * as MemoryAPI from "./memory.mjs";
import { Memory } from "./memory.mjs";
import * as ProceduresAPI from "./procedures.mjs";
import { Procedures } from "./procedures.mjs";
import * as SemanticMemoryAPI from "./semantic-memory.mjs";
import { SemanticMemory } from "./semantic-memory.mjs";
import { PageNumber } from "../../core/pagination.mjs";
export class Query extends APIResource {
    constructor() {
        super(...arguments);
        this.memory = new MemoryAPI.Memory(this._client);
        this.episodicMemory = new EpisodicMemoryAPI.EpisodicMemory(this._client);
        this.procedures = new ProceduresAPI.Procedures(this._client);
        this.semanticMemory = new SemanticMemoryAPI.SemanticMemory(this._client);
    }
    /**
     * Chunk Search provides search capabilities that serve as a complete drop-in
     * replacement for traditional RAG pipelines. This system enables AI agents to
     * leverage private data stored in SmartBuckets with zero additional configuration.
     *
     * Each input query is processed by our AI agent to determine the best way to
     * search the data. The system will then return the most relevant results from the
     * data ranked by relevance on the input query.
     *
     * @example
     * ```ts
     * const response = await client.query.chunkSearch({
     *   bucketLocations: [
     *     {
     *       bucket: {
     *         name: 'my-smartbucket',
     *         version: '01jxanr45haeswhay4n0q8340y',
     *         application_name: 'my-app',
     *       },
     *     },
     *   ],
     *   input: 'Find documents about revenue in Q4 2023',
     *   requestId: '<YOUR-REQUEST-ID>',
     * });
     * ```
     */
    chunkSearch(body, options) {
        return this._client.post('/v1/chunk_search', { body, ...options });
    }
    /**
     * Enables natural conversational interactions with documents stored in
     * SmartBuckets. This endpoint allows users to ask questions, request summaries,
     * and explore document content through an intuitive conversational interface. The
     * system understands context and can handle complex queries about document
     * contents.
     *
     * The query system maintains conversation context throught the request_id,
     * enabling follow-up questions and deep exploration of document content. It works
     * across all supported file types and automatically handles multi-page documents,
     * making complex file interaction as simple as having a conversation.
     *
     * The system will:
     *
     * - Maintain conversation history for context when using the same request_id
     * - Process questions against file content
     * - Generate contextual, relevant responses
     *
     * Document query is supported for all file types, including PDFs, images, and
     * audio files.
     *
     * @example
     * ```ts
     * const response = await client.query.documentQuery({
     *   bucketLocation: {
     *     bucket: {
     *       name: 'my-smartbucket',
     *       version: '01jxanr45haeswhay4n0q8340y',
     *       application_name: 'my-app',
     *     },
     *   },
     *   input: 'What are the key points in this document?',
     *   objectId: 'document.pdf',
     *   requestId: '<YOUR-REQUEST-ID>',
     * });
     * ```
     */
    documentQuery(body, options) {
        return this._client.post('/v1/document_query', { body, ...options });
    }
    /**
     * Retrieve additional pages from a previous search. This endpoint enables
     * navigation through large result sets while maintaining search context and result
     * relevance. Retrieving paginated results requires a valid request_id from a
     * previously completed search.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const liquidmetalV1alpha1TextResult of client.query.getPaginatedSearch(
     *   { page: 1, pageSize: 10, requestId: '<YOUR-REQUEST-ID>' },
     * )) {
     *   // ...
     * }
     * ```
     */
    getPaginatedSearch(body, options) {
        return this._client.getAPIList('/v1/search_get_page', (PageNumber), {
            body,
            method: 'post',
            ...options,
        });
    }
    /**
     * Primary search endpoint that provides advanced search capabilities across all
     * document types stored in SmartBuckets.
     *
     * Supports recursive object search within objects, enabling nested content search
     * like embedded images, text content, and personally identifiable information
     * (PII).
     *
     * The system supports complex queries like:
     *
     * - 'Show me documents containing credit card numbers or social security numbers'
     * - 'Find images of landscapes taken during sunset'
     * - 'Get documents mentioning revenue forecasts from Q4 2023'
     * - 'Find me all PDF documents that contain pictures of a cat'
     * - 'Find me all audio files that contain information about the weather in SF in
     *   2024'
     *
     * Key capabilities:
     *
     * - Natural language query understanding
     * - Content-based search across text, images, and audio
     * - Automatic PII detection
     * - Multi-modal search (text, images, audio)
     *
     * @example
     * ```ts
     * const response = await client.query.search({
     *   bucketLocations: [
     *     {
     *       bucket: {
     *         name: 'my-smartbucket',
     *         version: '01jxanr45haeswhay4n0q8340y',
     *         application_name: 'my-app',
     *       },
     *     },
     *   ],
     *   input: 'All my files',
     *   requestId: '<YOUR-REQUEST-ID>',
     * });
     * ```
     */
    search(body, options) {
        return this._client.post('/v1/search', { body, ...options });
    }
    /**
     * Generates intelligent summaries of search result pages, helping users quickly
     * understand large result sets without reading through every document. The system
     * analyzes the content of all results on a given page and generates a detailed
     * overview.
     *
     * The summary system:
     *
     * - Identifies key themes and topics
     * - Extracts important findings
     * - Highlights document relationships
     * - Provides content type distribution
     * - Summarizes metadata patterns
     *
     * This is particularly valuable when dealing with:
     *
     * - Large document collections
     * - Mixed content types
     * - Technical documentation
     * - Research materials
     *
     * @example
     * ```ts
     * const response = await client.query.sumarizePage({
     *   page: 1,
     *   pageSize: 10,
     *   requestId: '<YOUR-REQUEST-ID>',
     * });
     * ```
     */
    sumarizePage(body, options) {
        return this._client.post('/v1/summarize_page', { body, ...options });
    }
}
Query.Memory = Memory;
Query.EpisodicMemory = EpisodicMemory;
Query.Procedures = Procedures;
Query.SemanticMemory = SemanticMemory;
